src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -3827,62 +3821,172 @@ VOID MethodTableBuilder::AllocateFieldDescs()
-    sigBuilder.AppendData(nGenericArgs);
-    for (unsigned int typearg = 0; typearg < nGenericArgs; typearg++)
+    ClassLoader::ResolveTokenToTypeDefThrowing(module, typeDefOrRef, &pModuleOfTypeDef, &tkTypeDef);
+
+    // First check to see if the type is byref-like
+    if (pModuleOfTypeDef->GetCustomAttribute(tkTypeDef,
+        WellKnownAttribute::IsByRefLike,
+        NULL, NULL) == S_OK)
     {
-        sigBuilder.AppendElementType(ELEMENT_TYPE_VAR);
-        sigBuilder.AppendData(typearg);
+        BuildMethodTableThrowException(IDS_CLASSLOAD_BYREF_OR_BYREFLIKE_STATICFIELD);
     }
 
-    DWORD cFakeSig;
-    PCCOR_SIGNATURE pFakeSig = (PCCOR_SIGNATURE)sigBuilder.GetSignature(&cFakeSig);
+    mdToken tkTypeDefExtends;
+    IfFailThrow(pModuleOfTypeDef->GetMDImport()->GetTypeDefProps(tkTypeDef, NULL, &tkTypeDefExtends));
+    bool thisIsAnEnum = false;
 
-    PCCOR_SIGNATURE pFieldSig = pMemberSignature + 1; // skip the CALLCONV_FIELD
+    if (TypeFromToken(tkTypeDefExtends) == mdtTypeDef)
+    {
+        if (pModuleOfTypeDef->IsSystem())
+        {
+            if (tkTypeDefExtends == CoreLibBinder::GetClassIfExist(CLASS__ENUM)->GetCl())
+            {
+                thisIsAnEnum = true;
+            }
+        }
+    }
+    else
+    {
+        _ASSERTE(TypeFromToken(tkTypeDefExtends) == mdtTypeRef);
+        LPCSTR pszNameSpace;
+        LPCSTR pszClassName;
+        pModuleOfTypeDef->GetMDImport()->GetNameOfTypeRef(tkTypeDefExtends, &pszNameSpace, &pszClassName);
+        if (pszNameSpace != NULL && pszClassName != NULL && strcmp(pszNameSpace, "System") == 0 && strcmp(pszClassName, "Enum") == 0)
+        {
+            Module *pModuleOfSystemEnumType;
+            mdTypeDef tkTypeDefOfSystemEnumType;
 
-    return MetaSig::CompareElementType(pFakeSig, pFieldSig,
-                                       pFakeSig + cFakeSig,  pMemberSignature + cMemberSignature,
-                                       GetModule(), GetModule(),
-                                       NULL, NULL);
+            ClassLoader::ResolveTokenToTypeDefThrowing(pModuleOfTypeDef, tkTypeDefExtends, &pModuleOfSystemEnumType, &tkTypeDefOfSystemEnumType);
 
-}
+            if (pModuleOfSystemEnumType != NULL && pModuleOfSystemEnumType->IsSystem())
+            {
+                thisIsAnEnum = true;
+            }
+        }
+    }
 
-//*******************************************************************************
-//
-// Used pByValueClass cache to mark self-references
-//
-static BOOL IsSelfRef(MethodTable * pMT)
-{
-    return pMT == (MethodTable *)-1;
+    if (thisIsAnEnum)
+    {
+        // If this is an enum, we can return the type of the instance field
+        IMDInternalImport *pEnumModuleImport = pModuleOfTypeDef->GetMDImport();
+        HENUMInternalHolder hEnumField(pEnumModuleImport);
+        IfFailThrow(hEnumField.EnumInitNoThrow(mdtFieldDef, tkTypeDef));
+        mdFieldDef tkFieldDef;
+        while (hEnumField.EnumNext(&tkFieldDef))
```

Comment:
This is a bit confusing so just verifying my understanding here: The basetype of an enum is expressed by having an instance field on the concrete enum type which is of the basetype, i.e. an int64 enum will be `struct SomeEnum : System.Enum { private int64 value; }`, and you're analyzing that?